lizard_connector/connector.py

```diff
@@ -0,0 +1,342 @@
+# coding=utf-8
+"""
+Connector to Lizard api (e.g. https://demo.lizard.net/api/v2) for python.
+
+Includes:
+- Endpoints (Lizard api endoints, e.g. TimeseriesEndpoint)
+- Fetcher (http handling)
+- queryfunctions for special cases such as geographical queries and time
+related queries other queries can be input as a dictionary
+- parserfunctions to parse the json obtained from Endpoint queries
+"""
+
+import json
+import urllib.request
+import urllib.parse
+
+from lizard_connector.query import QueryDictionary
+
+
+class LizardApiTooManyResults(Exception):
+    pass
+
+
+class Fetcher(object):
+
+    def __init__(self, endpoint, base="https://demo.lizard.net",
+                 max_results=1000, username=None, password=None,
+                 all_pages=True):
+        """
+        :param max_results: maximum number of results allowed from one get.
+        :param username: lizard-api user name to log in. Without one no login
+            is used.
+        :param password: lizard-api password to log in. Without one no login
+            is used.
+        :param all_pages: when set to True, on get all pages are obtained.
+            When set to False only the first page is obtained on get.
+        """
+        self.all_pages = all_pages
+        self.max_results = max_results
+        self.next_url = None
+        self.count = None
+        self.username = username
+        self.password = password
+
+    def get(self, url):
+        """
+        GET a json from the api.
+        :param url: Lizard-api valid url.
+        :return: a dictionary of the 'results'-part of the api-response.
+        """
+        json_ = self.request(url)
+        self.count = json_.get('count')
+        self.next_url = json_.get('next')
+        json_ = json_.get('results', json_)
+        count = self.count if self.count else 0
+
+        if count > self.max_results:
+            raise LizardApiTooManyResults(
+                'Too many results: {} found, while max {} are accepted'.format(
+                json_.get('count', '?'), self.max_results)
+            )
+        if self.all_pages:
+            for extra_json in self:
+                json_.update(extra_json.get('results', extra_json))
+        return json_
+
+    def post(self, url, uuid, data):
+        """
+        POST data to the api.
+        :param url: Lizard-api valid endpoint url.
+        :param uuid: UUID of the object in the database you wish to store
+                     data to.
+        :param data: Dictionary with the data to post to the api
+        """
+        post_url = urllib.parse.urljoin(urllib.parse.urljoin(url, str(uuid)),
+                                        'data')
+        self.request(post_url, data=json.dumps(data))
+
+    def request(self, url, data=None):
```

Comment:
Ik heb de neiging dit een duidelijker naam te geven. `download()`, `perform_request()`. Een request op zich hoeft nog niet iets actiefs te zijn. Het is handig qua naam aan te geven dat hier daadwerkelijk iets gebeurd.